### PR TITLE
PMM-11302 Restores capabilities for operators

### PR DESCRIPTION
**Pull request by gen1us2k**

### CODE REVIEW

**Review by recharte — COMMENTED** [review]
As I was reviewing this PR a few things started to stand out as strange, inelegant implementation details that seemed necessary to make the restoration process work. e.g.
https://github.com/percona/dbaas-operator/blob/f848814c4120fea8f820a493c4d5ce4d1814fe7b/controllers/databaseclusterrestore_controller.go#L74-L76
https://github.com/percona/dbaas-operator/blob/f848814c4120fea8f820a493c4d5ce4d1814fe7b/controllers/databaseclusterrestore_controller.go#L305-L320
https://github.com/percona/dbaas-operator/blob/f848814c4120fea8f820a493c4d5ce4d1814fe7b/controllers/databaseclusterrestore_controller.go#L255-L287

This triggered me into thinking that maybe there's an underlying design issue here.

We are creating a `DatabaseClusterRestore` pseudo-type that at its core represents an action rather than a resource.
Of course, there's the `BackupSource` which is an integral part of this feature but I don't think it stands as a strong foundation for this type/controller. It definitely feels to me more like an input parameter than full blown k8s CR.

I might be a bit biased here due to my recent involvement in the templates design but I do see some resemblance. I see `BackupSource` having a role in the restoration process that's similar to the one that Templates have in the creation of a DBCluster, like an input param to an action. It seems to me that there shouldn't be any controller logic related to such a resource.
Maybe, the DBCluster controller should be the entity responsible for managing the restoration process with one (or more) new state(s) that identify the on-going restoration process. Until now we avoided diverging the DBCluster states from the ones followed by our operators but I think that this does indeed represent a new unique state.
Upon completion of the restoration process we can persist the restoration metadata (restore point, completion time, etc) as an annotation (or similar) on the DBCluster CR.
As for the bootstrap of this process, we could have a blend of the cluster restart process and the templates, i.e. denote the action you intend to do (restore) and reference some resource you require (`BackupSource`) in the form of annotations.

Having the DBCluster controller handle this reconciliation seems more native and avoids the aforementioned slightly hacky busy waiting required by the current implementation.

What are your thoughts?

File: controllers/databaseclusterrestore_controller.go [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+	cluster := &dbaasv1.DatabaseCluster{}
+	err = r.Get(ctx, types.NamespacedName{Name: cr.Spec.DatabaseCluster, Namespace: cr.Namespace}, cluster)
+	if err != nil {
+		logger.Error(err, "unable to get DatabaseCluster")
+		return reconcile.Result{}, err
+	}
+	orig := cluster.DeepCopy()
+
+	err = r.stopCluster(cluster.DeepCopy())
```
[the remark this answers is not in the document]
Comment (gen1us2k) [reply controllers/databaseclusterrestore_controller.go:86]: Yep. The idea is to restore everything that was present in `DatabaseCluster` before the restoration. stopCluster will change `Pause` field in the `DatabaseCluster` object and later the controller may restore a different version of the object

File: controllers/databaseclusterrestore_controller.go [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+	if err := r.Get(context.Background(), types.NamespacedName{Name: cluster.Name, Namespace: cluster.Namespace}, current); err != nil {
+		return err
+	}
+	current.Spec = cluster.Spec
+	current.Spec.Pause = false
+
+	if err := r.Update(context.Background(), current); err != nil {
+		return err
+	}
+	timeout, cancel := context.WithTimeout(context.Background(), 20*time.Minute)
```
[the remark this answers is not in the document]
Comment (gen1us2k) [reply controllers/databaseclusterrestore_controller.go:303]: Yeah. I agree. it was the first and naive implementation without following best practices

**Comment by gen1us2k** [comment]
@recharte Thank you for this comment

> We are creating a DatabaseClusterRestore pseudo-type that at its core represents an action rather than a resource.
> Of course, there's the BackupSource which is an integral part of this feature but I don't think it stands as a strong 
> foundation for this type/controller. It definitely feels to me more like an input parameter than full blown k8s CR.

For the 'Create a database cluster from backup' feature that we're working on right now, this applies very well. However, it's only the beginning of the journey for the database cluster restoration. What we have right now here is that 'on-demand restoration that is used to create a database cluster from a backup'. Later, the restoration feature will have its own fields that may differ from what we have in underlying operators. Also, we do not have on-demand backups yet. With them, it would be more natural for this. 

Having a controller for the `DatabaseClusterRestore` and having this particular kind allows us to implement the following features 

1. Decrypt secrets for the database cluster (It will be part of on-demand backups)
2. Implement `restore-and-swap` way of restoration against running database cluster which means it will follow these steps: spin-up a new cluster, clone secrets, run restoration, run the database cluster. Restore-and-swap is more secure way of restoration because there can be data left on the running cluster and dbaas-operator will prevent losing this data
3. It allows running on-demand restoration from backup against the running cluster while the database cluster is up and running. We do not have this feature yet for PMM/DBaaS but it opens a room to use it via k8s native way.


> I might be a bit biased here due to my recent involvement in the templates design but I do see some resemblance. I see 
> BackupSource having a role in the restoration process that's similar to the one that Templates have in the creation of a 
> DBCluster, like an input param to an action. It seems to me that there shouldn't be any controller logic related to such a 
> resource.

There are some resemblances exist but the difference between these features but having separate controller for restoration allows end user to restore cluster at any time. 

> Maybe, the DBCluster controller should be the entity responsible for managing the restoration process with one (or more) 
>  new state(s) that identify the on-going restoration process. Until now we avoided diverging the DBCluster states from the 
> ones followed by our operators but I think that this does indeed represent a new unique state.

> Upon completion of the restoration process we can persist the restoration metadata (restore point, completion time, etc) 
> as an annotation (or similar) on the DBCluster CR.
> As for the bootstrap of this process, we could have a blend of the cluster restart process and the templates, i.e. denote the 
> action you intend to do (restore) and reference some resource you require (BackupSource) in the form of annotations.

I like the idea, but we have these [best practices](https://sdk.operatorframework.io/docs/best-practices/best-practices/) for operators development

>An Operator should manage a single type of application, essentially following the UNIX principle: do one thing and do it well.

> Inside an Operator, multiple controllers should be used if multiple CRDs are managed. This helps in separation of concerns and code readability. Note that this doesn’t necessarily mean that we need to have one container image per controller, but rather one reconciliation loop (which could be running as part of the same Operator binary) per CRD.

In a nutshell, having a CR and a controller allows us

- Having more readable code 
- Having isolated logic between restoration and database cluster creation (We will have the same CR and controller for on-demand backups)
- Do not blow and overcomplicate the code of DatabaseCluster reconciliation loop

**Review by recharte — APPROVED** [review]
Much better without the starting and stopping instances trickery :clap:

File: controllers/databaseclusterrestore_controller.go [hunk 3]
```diff
@@ -86,7 +88,31 @@ func (r *DatabaseClusterRestoreReconciler) Reconcile(ctx context.Context, req ct
...
+			cluster := &dbaasv1.DatabaseCluster{}
+			err := r.Get(context.Background(), types.NamespacedName{Name: restore.Spec.DatabaseCluster, Namespace: restore.Namespace}, cluster)
+			if err != nil {
+				return err
+			}
+			if cluster.Status.State == dbaasv1.AppStateReady {
+				return nil
+			}
+		}
+	}
```
Comment (recharte) [inline controllers/databaseclusterrestore_controller.go:109]: What happens while we are busy waiting here? Are we able to reconcile other objects or is the entire DatabaseClusterRestoreReconciler blocked?
